### ISSUE TITLE
More tiny crates?

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3817]
We have a few problems right now that look like the solution is more tiny crates. This issue is to discuss if we want to do that. An alternative is to duplicate code and make things private.

1. `InputString`  and `writer_err!`: See also #3708 and #3322
2. `ParseIntError`: Uses `InputString`
3. Hex prefix string parsing functions and error types: Uses `ParseIntError`

These three things are all slightly unrelated and also exist for slightly different reasons. All three are used in multiple crates and are needed in order to stabalize.

- (1) exists solely to assist with error handling in a no-std environment
- (2) exists to add more context to integer parsing (includes multi-radix support)
- (3) exits to enable parsing integers from hex strings (only relates to hex)

### Where should they live

Currently

- (1) is in `internals`
- (2) and (3) are in `units`

Possible solutions:

- Make everything private and duplicate everywhere 
- Make 3 tiny crates
- Some mix of the two above
- Leave as is and muddy up the dependencies (make `hex` depend on `bitcoin-units`) 
- ?

### Other info

[parse_int](https://crates.io/crates/parse_int) is already taken and does not do the same thing as (2)

**Comment by tcharding** [comment]
How about a file/module in each crate that says at the top "this is an exact copy of the ones in crates x, y, and z" then we run diff in CI (warn only) to ensure they stay in sync. That means that any change can just be done to one then copy the whole file to the others?

**Comment by apoelstra** [comment]
I would like that. Arguably we should have a "master copy" somewhere with a script for copying it into place everywhere alongside the script for checking that the copy is faithful.

BTW I had suggested this some time ago and Kix complained that we were making up our own ad-hoc macro system and we should just use macros. I disagree with this claim (Rust macros are simultaneously too complicated to reason about, especially when crossing crate boundaries, and not powerful enough to usefully generate code) but we never resolved it. I think if he's not here we can go ahead and do it.

Though we should limit ourselves to just copying and requiring perfect diffs -- my original suggestion was also to add a pass with `cpp`/`php`/`nix`/`awk`/whatever so we could programmatically generate code :P. I can see why Kix would be opposed to that.

**Comment by tcharding** [comment]
`internals` is the first place that jumps to my mind to have the SSOT (Kix will at least like that we are using that concept).

**Comment by apoelstra** [comment]
Yeah, I'd be happy with that.

We may decide later that we want to move the files out of there into `ci/` or something but agreed, `internals` seems like a good starting place.

**Comment by tcharding** [comment]
It doesn't solve the problem of `hex` though. Kix previously pushed back on `hex` having a dependency on `bitcoin-internals`. Should `internals` move to its own repo and represent "all the shit that we opinionated rust-bitcoin devs think should be in, or done differently, in Rust. `io` would be a good candidate to move to said repo

**Comment by apoelstra** [comment]
Yeah, `io` is definitely user-facing and needs to be stable. `internals` has no stability guarantee.

Depending on what `hex` uses from `internals`, I wouldn't mind replacing the dependency with a copy of individual files from `internals`. We would want to be reasonably careful not to expose anything from there in the public API, since otherwise we'd be committed to it and couldn't undo it if we later found a nicer solution.

**Comment by tcharding** [comment]
> It doesn't solve the problem of `hex` though.

Oh boy, I was confused. There is no hex vs internals problem with this solution because we are copying the files - face palm.

So, to be sure we are on the same page, consensus in this thread is:

- Put all the "general purpose shit that should be part of Rust" in `internals` (types, macros, functions)
- Put each thing in a separate file.
- Copy each file from `internals` into any crate that uses needs them (see comment below)
- Add some tooling that diffs the files against `internals`
- Make sure none of these pieces is ever in the public API of any crate

We could even have a directory in `internals` called `private` and rsync it to every crate but edit the `private/mod.rs` file in each crate to only declare modules that are used. Then the other code is not built in. Then we just diff the whole directory in a single standard way for all crates. Docs explaining the whole setup could be in `mod.rs` as well.

**Comment by apoelstra** [comment]
>So, to be sure we are on the same page, consensus in this thread is:

Yes, though I think our current situation is fine in cases where:

* We are depending on `internals` for multiple reasons and/or we're willing to stomach an explicit `internals` dependency  for whatever reason
* We aren't calling macros from `internals` (or, we're calling macros that have no feature-gating or feature related stuff)

**Comment by tcharding** [comment]
Initial attempt in #3844

**Comment by tcharding** [comment]
`units` is done, this can be closed.
